pkg/gardenlet/controller/vpaevictionrequirements/add.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpaevictionrequirements
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+)
+
+// ControllerName is the name of this controller
+const ControllerName = "vpa-eviction-requirements"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(_ context.Context, mgr manager.Manager, seedCluster cluster.Cluster) error {
+	if r.SeedClient == nil {
+		r.SeedClient = seedCluster.GetClient()
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+
+	vpaEvictionRequirementsManagedByControllerPredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{
+		MatchLabels: map[string]string{constants.LabelVPAEvictionRequirementsController: "managed-by-controller"},
+	})
+	if err != nil {
+		return fmt.Errorf("failed computing label selector predicate for eviction requirements managed by controller: %w", err)
+	}
+
+	_, err = builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+		}).
+		WatchesRawSource(
+			source.Kind(seedCluster.GetCache(), &vpaautoscalingv1.VerticalPodAutoscaler{}),
+			&handler.EnqueueRequestForObject{}, builder.WithPredicates(
+				vpaEvictionRequirementsManagedByControllerPredicate,
+				predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update),
+				predicate.GenerationChangedPredicate{}),
+		).
+		Build(r)
+	return err
```

Comment:
Nit:

```suggestion
	return builder.
		ControllerManagedBy(mgr).
		Named(ControllerName).
		WithOptions(controller.Options{
			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
		}).
		WatchesRawSource(
			source.Kind(seedCluster.GetCache(), &vpaautoscalingv1.VerticalPodAutoscaler{}),
			&handler.EnqueueRequestForObject{}, builder.WithPredicates(
				vpaEvictionRequirementsManagedByControllerPredicate,
				predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update),
				predicate.GenerationChangedPredicate{},
			),
		).
		Complete(r)
```

Reply:
This is so neat! I didn't know this existed, thanks!